text-serializer-legacy/src/main/java/net/kyori/adventure/text/serializer/legacy/LegacyCodeConverter.java

```diff
@@ -0,0 +1,171 @@
+package net.kyori.adventure.text.serializer.legacy;
+
+import java.util.List;
+import net.kyori.adventure.text.format.NamedTextColor;
+import net.kyori.adventure.text.format.TextColor;
+import net.kyori.adventure.text.format.TextFormat;
+import net.kyori.adventure.util.HSVLike;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import static java.util.Objects.requireNonNull;
+
+public abstract class LegacyCodeConverter {
+  protected static final char LEGACY_BUNGEE_HEX_CHAR = 'x';
+
+  protected final char character;
+  protected final char hexCharacter;
+  protected final boolean hexColours;
+  protected final boolean useTerriblyStupidHexFormat; // (╯°□°)╯︵ ┻━┻
+
+  protected LegacyCodeConverter(final char character, final char hexCharacter, final boolean hexColours, final boolean useTerriblyStupidHexFormat) {
+    this.character = character;
+    this.hexCharacter = hexCharacter;
+    this.hexColours = hexColours;
+    this.useTerriblyStupidHexFormat = useTerriblyStupidHexFormat;
+  }
+
+  @NotNull abstract List<TextFormat> getFormats();
+
+  @NotNull abstract String getAllLegacyChars();
```

Comment:
In my opinion, instead of this we should have some sort of object containing this information:
```java
public final class CharacterAndFormat {
  private final char character;
  private final TextFormat format;
}
```
and pass a `List<CharacterAndFormat>` to the constructor which can convert those into the internal list/string representation.